cocotb/types/logic_array.py

```diff
@@ -0,0 +1,247 @@
+
+    @property
+    def integer(self) -> int:
+        value = 0
+        for bit in self:
+            value = value << 1 | int(bit)
+        return value
+
+    @property
+    def signed_integer(self) -> int:
+        value = self.integer
+        if value >= (1 << (len(self) - 1)):
+            value -= 1 << len(self)
+        return value
+
+    @typing.overload
+    def __setitem__(self, item: int, value: LogicConstructibleT) -> None:
+        ...
+
+    @typing.overload
+    def __setitem__(
+        self, item: slice, value: typing.Iterable[LogicConstructibleT]
+    ) -> None:
+        ...
+
+    def __setitem__(
+        self,
+        item: typing.Union[int, slice],
+        value: typing.Union[LogicConstructibleT, typing.Iterable[LogicConstructibleT]],
+    ) -> None:
+        if isinstance(item, int):
+            super().__setitem__(item, Logic(typing.cast(LogicConstructibleT, value)))
+        elif isinstance(item, slice):
+            super().__setitem__(
+                item,
+                (
+                    Logic(v)
+                    for v in typing.cast(typing.Iterable[LogicConstructibleT], value)
+                ),
+            )
+        else:
+            raise TypeError(
+                "indexes must be ints or slices, not {}".format(type(item).__name__)
+            )
+
+    def __repr__(self) -> str:
+        return "{}({!r}, {!r})".format(type(self).__qualname__, self.binstr, self.range)
+
+    def __and__(self: Self, other: Self) -> Self:
+        if isinstance(other, type(self)):
+            if len(self) != len(other):
+                raise ValueError(
+                    "cannot perform bitwise & on arrays of different length"
+                )
+            return type(self)(a & b for a, b in zip(self, other))  # type: ignore
+        return NotImplemented
+
+    def __rand__(self: Self, other: Self) -> Self:
+        return self & other
+
+    def __or__(self: Self, other: Self) -> Self:
+        if isinstance(other, type(self)):
+            if len(self) != len(other):
+                raise ValueError(
+                    "cannot perform bitwise | on arrays of different length"
+                )
+            return type(self)(a | b for a, b in zip(self, other))  # type: ignore
+        return NotImplemented
+
+    def __ror__(self: Self, other: Self) -> Self:
+        return self | other
+
+    def __xor__(self: Self, other: Self) -> Self:
+        if isinstance(other, type(self)):
+            if len(self) != len(other):
+                raise ValueError(
+                    "cannot perform bitwise ^ on arrays of different length"
+                )
+            return type(self)(a ^ b for a, b in zip(self, other))  # type: ignore
+        return NotImplemented
+
+    def __rxor__(self: Self, other: Self) -> Self:
+        return self ^ other
+
+    def __invert__(self: Self) -> Self:
+        return type(self)(~v for v in self)
+
+
+def _int_to_bitstr(value: int, bit_length: int) -> str:
```

Comment:
We have historically used `n_bits` to mean "bit length".
You are also using `bitlen` elsewhere, maybe normalize those two.